File: asion-base/src/main/java/org/asion/base/ddd/domain/BaseRepository.java
package org.asion.base.ddd.domain;

import org.asion.base.ddd.domain.annotations.DomainRepository;
import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.util.Optional;

/**
 * BaseRepository C R U D.
 *
 * @author Asion.
 * @since 2017/3/20.
 */
@DomainRepository
public interface BaseRepository<T extends BaseDomainEntity, ID extends Serializable> {


    /**
     * Saves a given entity. Use the returned instance for further operations as the save operation might have changed the
     * entity instance completely.
     *
     * @param entity must not be {@literal null}.
     * @return the saved entity will never be {@literal null}.
     */
    @NotNull
    <S extends T> S create(S entity);

    /**
     * Update entity
     *
     * @param entity must not be {@literal null}.
     * @return the updated entity will never be {@literal null}.
     */
    @NotNull
    <S extends T> S update(S entity);

    /**
     * Retrieves an entity by its id.
     *
     * @param id must not be {@literal null}.
     * @return the entity with the given id or {@literal Optional#empty()} if none found
     * @throws IllegalArgumentException if {@code id} is {@literal null}.
     */
    Optional<T> findById(ID id);

    /**
     * find one
     *
     * @param id id
     * @return entity
     */
    T findOne(ID id);

}

File: asion-bot/asion-bot-core/src/main/java/org/asion/bot/domain/model/BaseRepository.kt
package org.asion.bot.domain.model

import java.io.Serializable

/**
 * Comment Base Repository CRUD
 *
 * @author Asion
 * @since 2017-01-20
 */
interface BaseRepository<T, in ID : Serializable> {

    /**
     * 保存, 包括新增与更新
     *
     * @param entity 实体
     * @return 已经保存的实体
     */
    fun <S : T> save(entity: S): S

    /**
     * Saves all given entities.
     *
     * @param entities 实体
     * @return the saved entities
     * @throws IllegalArgumentException in case the given entity is null.
     */
    fun <S : T> save(entities: List<S>): List<S>

    /**
     * 根据主键返回
     *
     * @param id 主键
     * @return 对象
     */
    fun findOne(id: ID): T?

    /**
     * 根据ids返回实体列表
     * Returns the instances of the type with the given IDs.
     *
     * @param ids the given IDs.
     * @return the instances of the given IDs.
     */
    fun findList(ids: List<ID>): List<T>

    /**
     * 根据id返回对象是否存在
     *
     * @param id must not be null.
     * @return true if an entity with the given id exists, false otherwise
     */
    fun exists(id: ID): Boolean

    /**
     * 根据主键删除
     *
     * @param id 主键
     */
    fun delete(id: ID): Boolean

    /**
     * Deletes a given entity.
     *
     * @param entity Need to delete entity.
     * @throws IllegalArgumentException in case the given entity is null.
     */
    fun delete(entity: T): Boolean

}
